lib/std/sort.zig

```diff
@@ -502,6 +502,257 @@ test "binarySearch" {
+}
+
+/// Returns the index of the first element in `items` greater than or equal to `key`,
+/// or `items.len` if no such element exists.
+/// O(log n) complexity.
+pub fn upperBound(
+    comptime T: type,
+    key: anytype,
+    items: []const T,
+    context: anytype,
+    comptime lessThan: fn (context: @TypeOf(context), lhs: @TypeOf(key), rhs: T) bool,
+) usize {
+    var left: usize = 0;
+    var right: usize = items.len;
+    var mid: usize = undefined;
+
+    while (left < right) {
+        mid = (right + left) / 2;
+        if (!lessThan(context, key, items[mid])) {
+            left = mid + 1;
+        } else {
+            right = mid;
+        }
+    }
+
+    return left;
+}
+
+test "upperBound" {
+    const S = struct {
+        fn lower_u32(context: void, lhs: u32, rhs: u32) bool {
+            _ = context;
+            return lhs < rhs;
+        }
+        fn lower_i32(context: void, lhs: i32, rhs: i32) bool {
+            _ = context;
+            return lhs < rhs;
+        }
+        fn lower_f32(context: void, lhs: f32, rhs: f32) bool {
+            _ = context;
+            return lhs < rhs;
+        }
+    };
+
+    try testing.expectEqual(
+        @as(usize, 0),
+        upperBound(u32, @as(u32, 0), &[_]u32{}, {}, S.lower_u32),
+    );
+    try testing.expectEqual(
+        @as(usize, 0),
+        upperBound(u32, @as(u32, 0), &[_]u32{ 2, 4, 8, 16, 32, 64 }, {}, S.lower_u32),
+    );
+    try testing.expectEqual(
+        @as(usize, 1),
+        upperBound(u32, @as(u32, 2), &[_]u32{ 2, 4, 8, 16, 32, 64 }, {}, S.lower_u32),
+    );
+    try testing.expectEqual(
+        @as(usize, 2),
+        upperBound(u32, @as(u32, 5), &[_]u32{ 2, 4, 8, 16, 32, 64 }, {}, S.lower_u32),
+    );
+    try testing.expectEqual(
+        @as(usize, 6),
+        upperBound(u32, @as(u32, 8), &[_]u32{ 2, 4, 7, 7, 7, 7, 16, 32, 64 }, {}, S.lower_u32),
+    );
+    try testing.expectEqual(
+        @as(usize, 6),
+        upperBound(u32, @as(u32, 8), &[_]u32{ 2, 4, 8, 8, 8, 8, 16, 32, 64 }, {}, S.lower_u32),
+    );
+    try testing.expectEqual(
+        @as(usize, 3),
+        upperBound(u32, @as(u32, 8), &[_]u32{ 2, 4, 8, 16, 32, 64 }, {}, S.lower_u32),
+    );
+    try testing.expectEqual(
+        @as(usize, 6),
+        upperBound(u32, @as(u32, 64), &[_]u32{ 2, 4, 8, 16, 32, 64 }, {}, S.lower_u32),
+    );
+    try testing.expectEqual(
+        @as(usize, 6),
+        upperBound(u32, @as(u32, 100), &[_]u32{ 2, 4, 8, 16, 32, 64 }, {}, S.lower_u32),
+    );
+    try testing.expectEqual(
+        @as(usize, 2),
+        upperBound(i32, @as(i32, 5), &[_]i32{ 2, 4, 8, 16, 32, 64 }, {}, S.lower_i32),
+    );
+    try testing.expectEqual(
+        @as(usize, 1),
+        upperBound(f32, @as(f32, -33.4), &[_]f32{ -54.2, -26.7, 0.0, 56.55, 100.1, 322.0 }, {}, S.lower_f32),
+    );
+}
+
+/// Returns a tuple of the lower and upper bounds of `key` in `items`.
+/// See `lowerBound` and `upperBound` respectively.
+/// O(log n) complexity.
+pub fn equalRange(
```

Comment:
I think this name is misleading. I suggest 
```suggestion
pub fn binarySearchSpan(
```
and to change the doc comments to be more accurate and describe the edge cases.

Reply:
Will do on the doc comments 👍 

The name comes from C++ (1) and the following property:

```
const nums = [_]i32{ 2, 4, 8, 8, 8, 8, 15, 22 };
const x: i32 = 8;
const range = equalRange(i32, x, &nums, {}, S.lower_i32);
const all_equal = nums[range[0]..range[1]]; // { 8, 8, 8, 8 } - the range where all items are equal to `x`
```
Of course Zig is not C++, but I personally don't find the name misleading. Happy to change to `binarySearchSpan` if that is still your preference.

(1) https://en.cppreference.com/w/cpp/algorithm/equal_range

Reply:
Looking at the unit tests again, I see that it does in fact return a range with only equal elements, and an empty range if no elements are equal. I think "equal range" is fine.